modules/core/src/test/java/org/apache/ignite/cache/affinity/rendezvous/MdcAffinityBackupFilterSelfTest.java

```diff
@@ -0,0 +1,253 @@
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Verifies that {@link MdcAffinityBackupFilter} prohibits single data center deployment.
+     */
+    @Test
+    public void testSingleDcDeploymentIsProhibited() {
+        dcIds = new String[] {"DC_0"};
+        System.setProperty(IgniteSystemProperties.IGNITE_DATA_CENTER_ID, dcIds[0]);
+
+        backups = 1;
+        verifyGridFailsToStartWithMessage("Number of datacenters must be at least 2.");
+    }
+
+    /**
+     * Verifies that {@link MdcAffinityBackupFilter} enforces even number of partition copies per datacenter.
+     */
+    @Test
+    public void testEvenNumberOfPartitionCopiesPerDcIsEnforced() {
+        dcIds = new String[] {"DC_0", "DC_1", "DC_2"};
+        System.setProperty(IgniteSystemProperties.IGNITE_DATA_CENTER_ID, dcIds[0]);
+
+        backups = 1;
+        verifyGridFailsToStartWithMessage("recommended value is 2");
+
+        backups = 7;
+        verifyGridFailsToStartWithMessage("recommended values are 5 and 8");
+    }
+
+    /**
+     * Verifies that partition copies are assigned evenly across a cluster in two data centers.
+     * <p/>
+     * When a node from one data center is stopped, partition distribution is that data center should stay uniform.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void test2DcDistribution() throws Exception {
+        dcIds = new String[] {"DC_0", "DC_1"};
+        int nodesPerDc = 4;
+        backups = 3;
+
+        IgniteEx srv = startClusterAcrossDataCenters(dcIds, nodesPerDc);
+
+        verifyDistributionProperties(srv, dcIds, nodesPerDc, backups);
+
+        //stopping one node in DC_1 should not compromise distribution as there are additional nodes in the same DC
+        stopGrid(5);
+
+        awaitPartitionMapExchange();
+
+        verifyDistributionProperties(srv, dcIds, nodesPerDc, backups);
+
+        //stopping another node in DC_1 should not compromise distribution as well
+        stopGrid(6);
+
+        awaitPartitionMapExchange();
+
+        verifyDistributionProperties(srv, dcIds, nodesPerDc, backups);
+    }
+
+    /**
+     * Verifies that partition copies are assigned evenly across a cluster in three data centers.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void test3DcDistribution() throws Exception {
+        dcIds = new String[] {"DC_0", "DC_1", "DC_2"};
+        int nodesPerDc = 2;
+        backups = 5;
+
+        IgniteEx srv = startClusterAcrossDataCenters(dcIds, 2);
+
+        verifyDistributionProperties(srv, dcIds, nodesPerDc, backups);
+    }
+
+    /** */
+    private void verifyGridFailsToStartWithMessage(String msg) {
+        try {
+            startGrid(0);
+        } catch (IllegalArgumentException argEx) {
```

Comment:
`assertThrowsWithCause/assertThrows` can be used.

Reply:
Fixed